cache/manager.go

```diff
@@ -367,23 +404,58 @@ func (cm *cacheManager) prune(ctx context.Context, ch chan client.UsageInfo, fil
 				Shared:     shared,
 			}
 
-			if filter.Match(adaptUsageInfo(c)) {
-				cr.dead = true
+			usageCount, lastUsedAt := getLastUsed(cr.md)
+			c.LastUsedAt = lastUsedAt
+			c.UsageCount = usageCount
 
-				toDelete = append(toDelete, cr)
-
-				// mark metadata as deleted in case we crash before cleanup finished
-				if err := setDeleted(cr.md); err != nil {
+			if opt.keepDuration != 0 {
+				if lastUsedAt != nil && lastUsedAt.After(cutOff) {
 					cr.mu.Unlock()
-					cm.mu.Unlock()
-					return err
+					continue
 				}
 			}
-		}
 
+			if opt.filter.Match(adaptUsageInfo(c)) {
+				toDelete = append(toDelete, &deleteRecord{
+					cacheRecord: cr,
+					lastUsedAt:  c.LastUsedAt,
+					usageCount:  c.UsageCount,
+				})
+				if !gcMode {
+					cr.dead = true
+
+					// mark metadata as deleted in case we crash before cleanup finished
+					if err := setDeleted(cr.md); err != nil {
+						cr.mu.Unlock()
+						cm.mu.Unlock()
+						return err
+					}
+				} else {
+					locked[cr] = struct{}{}
+					continue // leave the record locked
+				}
+			}
+		}
 		cr.mu.Unlock()
 	}
 
+	if gcMode && len(toDelete) > 0 {
+		sortDeleteRecords(toDelete)
+		var err error
+		for i, cr := range toDelete {
+			if i == 0 {
```

Comment:
This `if i == 0` special case is not obvious to me. Maybe you could add a comment explaining?

Reply:
We only remove single record at a time and then check the conditions again